ksqldb-engine/src/main/java/io/confluent/ksql/function/udaf/array/CollectListUdaf.java

```diff
@@ -15,98 +15,105 @@
-    return new TableUdaf<T, List<T>, List<T>>() {
-
-      @Override
-      public List<T> initialize() {
-        return Lists.newArrayList();
-      }
-
-      @Override
-      public List<T> aggregate(final T thisValue, final List<T> aggregate) {
-        if (aggregate.size() < LIMIT) {
-          aggregate.add(thisValue);
-        }
-        return aggregate;
-      }
-
-      @Override
-      public List<T> merge(final List<T> aggOne, final List<T> aggTwo) {
-        final int remainingCapacity = LIMIT - aggOne.size();
-        aggOne.addAll(aggTwo.subList(0, Math.min(remainingCapacity, aggTwo.size())));
-        return aggOne;
-      }
-
-      @Override
-      public List<T> map(final List<T> agg) {
-        return agg;
-      }
-
-      @Override
-      public List<T> undo(final T valueToUndo, final List<T> aggregateValue) {
-        // A more ideal solution would remove the value which corresponded to the original insertion
-        // but keeping track of that is more complex so we just remove the last value for now.
-        final int lastIndex = aggregateValue.lastIndexOf(valueToUndo);
-        // If we cannot find the value, that means that we hit the limit and never inserted it, so
-        // just return.
-        if (lastIndex < 0) {
-          return aggregateValue;
-        }
-        aggregateValue.remove(lastIndex);
-        return aggregateValue;
-      }
-    };
-  }
-
   @UdafFactory(description = "collect values of a Bigint field into a single Array")
   public static TableUdaf<Long, List<Long>, List<Long>> createCollectListLong() {
-    return listCollector();
+    return new Collect<>();
   }
 
   @UdafFactory(description = "collect values of an Integer field into a single Array")
   public static TableUdaf<Integer, List<Integer>, List<Integer>> createCollectListInt() {
-    return listCollector();
+    return new Collect<>();
   }
 
   @UdafFactory(description = "collect values of a Double field into a single Array")
   public static TableUdaf<Double, List<Double>, List<Double>> createCollectListDouble() {
-    return listCollector();
+    return new Collect<>();
   }
 
   @UdafFactory(description = "collect values of a String/Varchar field into a single Array")
   public static TableUdaf<String, List<String>, List<String>> createCollectListString() {
-    return listCollector();
+    return new Collect<>();
   }
 
   @UdafFactory(description = "collect values of a Boolean field into a single Array")
   public static TableUdaf<Boolean, List<Boolean>, List<Boolean>> createCollectListBool() {
-    return listCollector();
+    return new Collect<>();
+  }
+
+  private static final class Collect<T> implements TableUdaf<T, List<T>, List<T>>, Configurable {
+
+    private int limit = Integer.MAX_VALUE;
+
+    @Override
+    public void configure(final Map<String, ?> map) {
+      final Object limit = map.get(LIMIT_CONFIG);
+      this.limit = (limit == null) ? this.limit : ((Number) limit).intValue();
```

Comment:
How are negative values handled?

Reply:
good catch, I will treat negative as "no limit". I'll add that to the documentation